cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,268 @@
+call cclutAddMockData("person", "4.0|Madison||04-APR-1973 10:33") ;Will add Madison (empty string for first name)
+
+**cclutClearMockData(tableName = vc)**
+
+Clears all data from the mock table.  This is functionally similar to a truncate.  tableName is required.  The table  
+must have been created through cclutCreateMockTable() or else an error will be thrown.  
+  
+@param tableName  
+&nbsp;&nbsp;&nbsp;&nbsp;The table to be cleared.  
+  
+Example:  
+call cclutClearMockData("person")
+
+**cclutAddMockImplementation(originalName = vc, replaceName = vc)**
+
+Adds a mock implementation to be utilized by cclutExecuteProgramWithMocks.  This can include record structures, subroutines, or other programs.  originalName and replaceName are required.  If originalName is already being mocked, the mock will be overwritten with the new replaceName.  
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object to be mocked.  
+@param replaceName  
+&nbsp;&nbsp;&nbsp;&nbsp;The mocked object.  
+  
+Example:  
+call cclutAddMockImplementation("uar_get_code_by", "mock_uar_get_code_by")
+
+**cclutRemoveMockImplementation(originalName = vc)**
+
+Removes a mock implementation.  
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object that is mocked.  
+  
+Example:  
+call cclutRemoveMockImplementation("uar_get_code_by")
+
+**cclutRemoveAllMockImplementations(null)**
+
+Removes all mock implementations.  
+  
+Example:  
+call cclutRemoveAllMockImplementations(null)
+
+**cclutExecuteProgramWithMocks(programName = vc, params = vc, namespace = vc)**
+
+Executes a program with all mocks currently added through cclutAddMockImplementation() and cclutCreateMockTable(). programName is required.  params is a string parameter to be sent directly to the program, so all commas and string delimiters must be specified.  If namespace is omitted, it will default to the PUBLIC namespace.  
+  
+Example:  
+call cclutExecuteProgramWithMocks("ccl_my_program", "\^MINE^, 1.0, ^string parameter^", "MYNAMESPACE")
+
+**cclutRemoveAllMocks**
+
+Removes all mock implementations and mock tables that have been added through the cclutAddMockImplementation() and cclutCreateMockTable() APIs.  This should be called at the completion of a test suite to clean up all mocks.  
+  
+Example:  
+call cclutRemoveAllMocks(null)
+
+## Implementation Notes
+1. cclutRemoveAllMocks should be called as part of the teardown for all tests.  The framework will attempt to clean up any outstanding mocks, but it is good practice to explicitly remove any mocks to ensure that no mocked tables remain in the Oracle instance.
+
+2. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to children script called from the script-under-test.  Some alternatives would be to mock the child script to return the appropriate data or to mock the child script to execute the real script applying the mocked tables and implementations.
```

Comment:
Well, I don't like the idea of bumping the minimum CCL version (since it's the absolute newest version that has the new "replace" functionality) when it would just be a side piece of functionality.  

We _could_ (I suppose) stipulate that the propagation functionality requires 8.14.3/9.02.3, while everything else is lower, but that feels like it creates a bad precedent, and we should stick to a singular version for everything.

So I'm waffling between not offering it in this initial release or doing the version check.  The version check might also lead to bad precedents (Why not do version checking for all features?), so I think I'm leaning more towards not offering it in the initial release, and then we can add it if there's a strong appetite from consumers.  Since it is new functionality to CCL, anyone that was doing mocking prior was already mocking their external calls (at least if they were honoring the spirit of a "unit" test), so I don't think it would be an undue burden to continue that.  It also reduces the complexity of this initial version as I want to ensure we have a strong foundation rather than trying to get every piece of functionality out the door in one release.

If you think it's important enough though, I can throw in the version check.

Reply:
As noted, I think such external calls should be mocked making it unnecessary so I would be inclined to wait for popular demand..... but think about not making it too difficult to add if heavily requested.

Reply:
Sounds good.  Understood.